finality-provider/service/fp_instance.go

```diff
@@ -717,6 +718,126 @@ func (fp *FinalityProviderInstance) SubmitBatchFinalitySignatures(blocks []*type
 	return res, nil
 }
 
+// SubmitBatchFinalitySignaturesBoth builds and sends a finality signature over the given block to the consumer chain
+// NOTE: the input blocks should be in the ascending order of height
+// NOTE: it will try EOTS signing generated through safe and unsafe versions of rand generator (if either one succeeds)
+// this is for backward compatibility, will be deprecated soon
+func (fp *FinalityProviderInstance) SubmitBatchFinalitySignaturesBoth(blocks []*types.BlockInfo) (*types.TxResponse, error) {
+	if len(blocks) == 0 {
+		return nil, fmt.Errorf("should not submit batch finality signature with zero block")
+	}
+
+	if len(blocks) > math.MaxUint32 {
+		return nil, fmt.Errorf("should not submit batch finality signature with too many blocks")
+	}
+
+	// sign block and send one-by-one
+	// we cannot send blocks in a batch in case one failure will cause the whole batch fail
+	var res *types.TxResponse
+	for _, b := range blocks {
+		proofBytes, err := fp.pubRandState.getPubRandProof(
+			fp.btcPk.MustMarshal(),
+			fp.GetChainID(),
+			b.Height,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get public randomness inclusion proof list: %w\nplease recover the randomness proof from db", err)
+		}
+
+		res, err = fp.trySendTwoVersionFinalitySigs(b, proofBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		// update DB
+		fp.MustUpdateStateAfterFinalitySigSubmission(b.Height)
+	}
+
+	return res, nil
+}
+
+// trySendTwoVersionFinalitySigs attempts to send finality signatures using both unsafe and safe versions
+// Returns the transaction response and any error encountered
+func (fp *FinalityProviderInstance) trySendTwoVersionFinalitySigs(block *types.BlockInfo, proof []byte) (*types.TxResponse, error) {
+	// Get message to sign
+	msgToSign := getMsgToSignForVote(block.Height, block.Hash)
+
+	// Generate both signature versions
+	sigRecord, sigRecordUnsafe, err := fp.em.SignEOTSBoth(
+		fp.btcPk.MustMarshal(),
+		fp.GetChainID(),
+		msgToSign,
+		block.Height,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate EOTS signatures: %w", err)
+	}
+
+	// Try unsafe version first
+	res, err := fp.trySignatureVersion(block, sigRecordUnsafe.PubRand, proof, sigRecordUnsafe.Sig, "unsafe")
+	if err == nil {
+		return res, nil
+	}
+
+	res, err = fp.trySignatureVersion(block, sigRecord.PubRand, proof, sigRecord.Sig, "safe")
+	if err != nil {
+		return nil, fmt.Errorf("both signature versions failed: %w", err)
+	}
```

Comment:
Is it possible to implement something as follows:

- query the pub rand at that height
- get the sign record with the same pub rand
- only submit the finality sig in that sign record

This way we avoid the retry completely

Reply:
Good point. I just realized that we can use the proof to verify which rand to use without any query